### DISCUSSION TITLE
Cache misses with same checkout in different locations, sloppiness and base dir

### DISCUSSION THREAD

**Discussion by timblechmann** [ccache/ccache#212] [Support]
i'm using ccache on a buildservers to speed up our builds. however i'm seen a weird issue:

* i'm setting the base dir via an environment variable to `Config: (environment) base_dir = /var/lib/bamboo/bamboo-agent-home/xml-data/build-dir`
* likewise i'm setting the sloppiness: `Config: (environment) sloppiness = file_macro, time_macros`
* the sources are checked out to: `/var/lib/bamboo/bamboo-agent-home/xml-data/build-dir/XXX/sources` and compiled in the build folder `/var/lib/bamboo/bamboo-agent-home/xml-data/build-dir/XXX/build`. where XXX is a unique identifier for each git branch that is i'm going to test.

when for each build in a branch, i have the expected cache hits. but in the first build of a branch (`XXX` is a new unique symbol), i'm seen cache misses.

if i consult the `*.ii` file which is generated in `.ccache/tmp`, the first 2 lines surprised me:
```
# 1 "../sources/path/to/my.cpp"
# 1 "/var/lib/bamboo/bamboo-agent-home/xml-data/build-dir/XXX/build//"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 1 "<command-line>" 2
# 1 "../sources/path/to/my.cpp"
```

i'im not sure if the `.ii` is the one that is taken into account for the hashing, but it contains a path of the build folder, probably coming from the current working directory.

other than that, are there any other possible explanation for the results that i'm seeing?

**Comment by jrosdahl** [comment]
Most likely, you either need to set `hash_dir = false` or start using `-fdebug-prefix-map`. See the documentation of `hash_dir` in the [_Configuration_ section of the manual](https://ccache.samba.org/manual/3.3.5.html#_configuration).

I now realize that this information should be present under [_Compiling in different directories_](https://ccache.samba.org/manual/latest.html#_compiling_in_different_directories) as well. I'll make it so in the next ccache release.

Please try it out and report back if it worked! :slightly_smiling_face:

**Comment by timblechmann** [comment]
ah ... that's it! `hash_dir` helps ... 

`-fdebug-prefix-map` doesn't seem to have an effect, though, but maybe i'm using it in a wrong way by setting it to `-fdebug-prefix-map='${CMAKE_SOURCE_DIR}'=.`

> Exception: The CWD will not be included in the hash if base_dir is set (and matches the CWD) and the compiler option -fdebug-prefix-map is used.

does "matches" refer to "is a subfolder of" or "equals"?

so, having a base dir of `/var/lib/bamboo/bamboo-agent-home/xml-data/build-dir/` sources in `/var/lib/bamboo/bamboo-agent-home/xml-data/build-dir/XXX/sources` and builds in `/var/lib/bamboo/bamboo-agent-home/xml-data/build-dir/XXX/build` would i have to set the debug prefix to `/var/lib/bamboo/bamboo-agent-home/xml-data/build-dir/` or to `/var/lib/bamboo/bamboo-agent-home/xml-data/build-dir/XXX/sources`?

**Comment by afbjorklund** [comment]
Maybe it doesn't account for relative paths apropriately ? Does it help if you use absolute paths ?

**Comment by afbjorklund** [comment]
Are you sure you want to use single quotes here: `-fdebug-prefix-map='${CMAKE_SOURCE_DIR}'=.`

**Comment by timblechmann** [comment]
the results usin `-fdebug-prefix-map=${CMAKE_SOURCE_DIR}=.` are ... funny (`${CMAKE_SOURCE_DIR}` boils down to `base_dir/XXX/sources`, while the build folder is `base_dir/XXX/build`)

* compiling sources in the source directory, i'm seeing a good number of cache hits.
* compiling generated sources in the build directory, i'm not getting any cache hits. this is a bit unfortunate, as i'm using *lots* of generated sources for unity builds (`#include`ing all cpp files into a single translation unit).

**Comment by afbjorklund** [comment]
The theory behind debug-prefix-map (and hash-dir), is that we want to rewrite the stuff that ends up in the object files so that we can re-use the cache files without causing issues for debugging (i.e. stale paths).
So if you have sources in both places (both original and generated), then you need to remap both locations. But there should be support for multiple `-fdebug-prefix-map` parameters these days, since 13341d006f4c8086d7d01cce96d48c8202d0c0aa ?

**Comment by afbjorklund** [comment]
Not sure how/if debugging actually works for your concatenated files, but maybe you sprinkle them with  `#line` or somehing ? Since otherwise debuginfo will link to a non-existing (generated) source file anyway.

**Comment by timblechmann** [comment]
thanks for the explanations. remapping both directories sounds logical, but unfortunately i'm on an older ccache version (3.3.4) and not hashing the working directory is a good workaround for me

adding these points to the "Compiling in different directories" would be much appreciated, though (that's where i was looking first and missed that point)

**Comment by jrosdahl** [comment]
> adding these points to the "Compiling in different directories" would be much appreciated, though (that's where i was looking first and missed that point)

Done: https://ccache.samba.org/manual/3.3.6.html#_compiling_in_different_directories

Does it look OK?

**Comment by timblechmann** [comment]
thanks a lot, looks good to me.
